### PR TITLE
Mitchell/jamie blog post

### PR DESCRIPTION
**Pull request by mitchelljohnson3**
Mitchell Johnson and Jamie Schmidt's blog post.

A summary of the work we completed this summer
How to use the code we completed.

### CODE REVIEW

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+---
+layout: topic
+title: "Development of MABE2"
+date: 2021-07-31
+author: Mitchell Johnson & Jamie Schmidt
+---
+
+### Development of MABE2
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:8]: The title will be automatically generated by Jekyll so you don’t have to put it in manually

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+## Introduction
+-what is MABE2
+For our project, we focused our work mostly on 
+-improving the end user experience through the development of DynamicOrg.
+-adding support for brains and genomes, both of which have base classes
+
+To quote the [Github for MABE2](https://github.com/mercere99/MABE2), MABE is a software framework deigned to easily build and customize software for evolutionary computation or artificial life.  The resulting systems should be useful for studying evolutionary dynamics, solving complex problems, comparing evolving systems, or exploring the open-ended power of evolution.
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:16]: You can use 
> block quotes
To do big quotes
And such

```
> block quotes
To do big quotes
And such
```

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+To quote the [Github for MABE2](https://github.com/mercere99/MABE2), MABE is a software framework deigned to easily build and customize software for evolutionary computation or artificial life.  The resulting systems should be useful for studying evolutionary dynamics, solving complex problems, comparing evolving systems, or exploring the open-ended power of evolution.  
+
+MABE, accessible [here](https://github.com/Hintzelab/MABE/wiki), stands for Modular Agent-Based Evolution and was one of the first true digital evolution frameworks.  It was structured in this fashion:
+
+![MABE structure](https://github.com/Hintzelab/MABE/wiki/images/MABE_Overview.png)
+
+and, though powerful, could of course be improved.  The creation of the 2.0 version of MABE, called MABE2, was the focus of our project.  Our main focus was on refining the underlying architecture of MABE to be more end user-friendly, particularly in designing custom experiments.  The culmination of our efforts was in the creation of DynamicOrg, an Organism which was near-completely abstracted to allow for any combination of Genomes and Brains, as assigned through the config file.
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:22]: You might consider incorporating schematics to show how different conceptual pieces fit together

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+![MABE structure](https://github.com/Hintzelab/MABE/wiki/images/MABE_Overview.png)
+
+and, though powerful, could of course be improved.  The creation of the 2.0 version of MABE, called MABE2, was the focus of our project.  Our main focus was on refining the underlying architecture of MABE to be more end user-friendly, particularly in designing custom experiments.  The culmination of our efforts was in the creation of DynamicOrg, an Organism which was near-completely abstracted to allow for any combination of Genomes and Brains, as assigned through the config file. 
+
+## Genomes / author: jamie
+Genomes in MABE2 are just repositories of mutatable information.  The most basic kinds are vectors of some type, and (unlike in MABE), they aren't hidden inside Organisms which manage everything about them, but they are actively added to them and manage themselves.  This means that to mutate a Genome, any Genome, the Organism just tells the Genome to mutate itself and the Genome does.  No longer will the Organism run all the mutation functions itself, now it just delegates.  This, critically, means that Organisms now no longer need to know what kinds of Genomes they have at all.
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:25]: Dig deeper into what the motivations/advantages are. What types of organisms or what types of savings of developer time are possible?

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+Now, a DynamicOrg must necessarily be significantly less restrictive.  Copying itself is fine, we just copy all its Genomes and Brains.  Mutations are also reasonably simple, since all we need to do is mutate every Genome in it.  But how on earth do we know what to return?  Simply put, we don't, so we just want to return an emp::DataMap.  The Evaluator is what knows, so we let the Evalutor decide.  The Brains then all edit this emp::DataMap in order, in whatever way they are designed to do.  
+  
+In the config file, this means all a DynamicOrg needs is a list of names of all the Genomes and Brains it will contain.  So long as all these Genomes and Brains are properly named in the confic file, this will work seamlessly. 
+  
+## Brains / author: jamie
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:50]: If you want to have authorship attributed for different parts, a kind of new thing that people are doing are “Author Contribution Statements” like this https://www.epj.org/images/stories/faq/examples-of-author-contributions.pdf You could just slot that in at the end

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  
+  1. The config system for Brains and Genomes does not yet work.  The structure should be similar to how Organisms are, with each Genome having a name and the various pertinent values (such as mutation rate and min/max values) and each Brain having a name, the names of the Genomes it will use, and whatever stats it will have (such as Markov Brains needing the size of their gates and all Brains having the name of their outputs).  
+  2. Brains should be implemented to actually use this information.
+  3. BasicBrains, MarkovBrains, and TypedGenomes should all be put in their own files.
+  4. The MarkovBrain should be properly tested and the hardcoded values should be made easily configurable via the config file.
+  5. As part of item 4, GenomeManager and BrainManager classes should be created to act as OrganismManagers do for Genomes and Brains.  Ideally, these would all be turned into one big Manager class.
+  6. The more complicated mutation types should be added to TypedGenome with configuration options to set the rate for each of them.
+  7. More Brain types should be added by default.
+  
+
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:118]: Don’t forget to add an acknowledgements section!

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  
+An Evaluator is, perhaps, the thing most users will spend the most time with for their custom experiments.  In any evolutionary situation, Organisms will need to be evaluated on whatever traits they are supposed to be evolving, and the Evaluator assigns their fitness value from whatever outputs the Brains provide.  
+  
+When coding an Evaluator, the most important thing is to make sure that, for whatever inputs your evaluate function requires, you've got Brains that can supply it.  They're configured in the same way that Brains and Genomes are, by being given a list of input names that correspond to the types that the Evaluator is expecting.
+  
+We did not make many changes to the Evaluator model, aside from making it compatible with DynamicOrgs, which in turn made them significantly easier to customize since every DynamicOrg Evaluator follows the same steps: 
+  1. Get the emp::DataMap
+  2. Get all the input values you want from it
+  3. Calculate fitness from them
+  4. Return fitness
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:72]: Each one of these sections is rich enough to be a post on its own! In terms of organizational flow, I think you’ll have to be careful that readers don’t lose their sense of the overall flow of the blog in the middle as they’re working through all this great detail. One solution might be to put a kind of table of contents with links into each section’s anchor headings and descriptions of what’s covered in each section (“tell them what you’re going to tell them” kind of situation). Or you could write a three or four sentence overview for each section in the main body blog post and put the full complete details on the end of the post as Appendix sections (just regular sections labeled appendix) that you could link to from the body.

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+## Tutorial - creating and running Aagos using DynamicOrg / author: mitchell
+-explanation of aagos and how it works
+-writing the evaluator
+-writing the aagos brain
+-putting it all together in the .gen file -> .mabe file
+-running the expiriment
+  
+  1. Write the Evaluator. (almost always necessary)
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:81]: You could set up a repl.it that people can try this in.

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      - Set up the config.
+      - Ensure compatibility with the base class.
+      - Customize the Mutate() function.
+  4. Write the Organism. (shouldn't be necessary except in the direst of circumstances)
+      - Set up the config.
+      - Customize the Initialize() and Mutate() functions.
+    
+
+## Conclusion
+-reiterate how our work greatly simplified the end user expirience using mabe2
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:99]: I saw in Mitchell’s abstract some specific quantified details about how much the process is being simplified, that would be great to include!!!

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+## Future plans / Work we didn't complete
+-anything we wish we could have completed but didn't have time to
+  
+Unfortunately, we simply did not have the time to do everything we wanted.  As such, these are the remaining issues:
+  
+  1. The config system for Brains and Genomes does not yet work.  The structure should be similar to how Organisms are, with each Genome having a name and the various pertinent values (such as mutation rate and min/max values) and each Brain having a name, the names of the Genomes it will use, and whatever stats it will have (such as Markov Brains needing the size of their gates and all Brains having the name of their outputs).  
+  2. Brains should be implemented to actually use this information.
+  3. BasicBrains, MarkovBrains, and TypedGenomes should all be put in their own files.
+  4. The MarkovBrain should be properly tested and the hardcoded values should be made easily configurable via the config file.
+  5. As part of item 4, GenomeManager and BrainManager classes should be created to act as OrganismManagers do for Genomes and Brains.  Ideally, these would all be turned into one big Manager class.
```
Comment (mmore500) [inline blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md:114]: This is a great overview! Try to add more intuition for why/how you think these would be fruitful approaches. Sometimes I like to think like I’m writing something for myself to read right before I got started on the project — what would I most want to communicate with myself and what would I need to hear to skip over months of figuring it out and end up thinking like I am now?

**Comment by mmore500** [comment]
@mitchelljohnson3 @schmi710 also, please also date the post for 08-20 (the last day of the workshop)

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md (hunk 1 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md]: removed

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md (hunk 3 above)
Comment (schmi710) [reply blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md]: Will do!

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md (hunk 4 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md]: done

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md (hunk 10 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md]: good idea, added

File: blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md (hunk 6 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-07-31-Mitchell-Johnson_Jamie-Schmidt.md]: added!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+layout: post
+date: 2021-07-31
+author: Mitchell Johnson & Jamie Schmidt
+---
+
+### Development of MABE2
+
+## Introduction
+
+To quote the [Github for MABE2](https://github.com/mercere99/MABE2), MABE is a software framework deigned to easily build and customize software for evolutionary computation or artificial life.  The resulting systems should be useful for studying evolutionary dynamics, solving complex problems, comparing evolving systems, or exploring the open-ended power of evolution.
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:11]: add " "

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+## Introduction
+
+To quote the [Github for MABE2](https://github.com/mercere99/MABE2), MABE is a software framework deigned to easily build and customize software for evolutionary computation or artificial life.  The resulting systems should be useful for studying evolutionary dynamics, solving complex problems, comparing evolving systems, or exploring the open-ended power of evolution.  
+
+MABE, accessible [here](https://github.com/Hintzelab/MABE/wiki), stands for Modular Agent-Based Evolution and was one of the first true digital evolution frameworks.  It was structured in this fashion:
+
+![MABE structure](https://github.com/Hintzelab/MABE/wiki/images/MABE_Overview.png)
+
+and, though powerful, could of course be improved.  The creation of the 2.0 version of MABE, called MABE2, was the focus of our project.  Our main focus was on refining the underlying architecture of MABE to be more end user-friendly, particularly in designing custom experiments.  The culmination of our efforts was in the creation of DynamicOrg, an Organism which was near-completely abstracted to allow for any combination of Genomes and Brains, as assigned through the config file.
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:17]: change "and, though" to "Though" since the start of that sentence is separated by the image and this could be a stand alone sentence

Comment (AriaKillebrewBruehl) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:17]: bold "The culmination..." since that's the main point of all your work and that sentence gets a little lost in the paragrpah

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+  
+For comparison, this is the structure of MABE2 when using a DynamicOrg:
+  
+![MABE2 structure](https://i.imgur.com/TCklKle.png)
+  
+To explain exactly what's going on here, we're going to go through Genomes, DynamicOrg, Brains, and then Evaluators.
+
+## Genomes / author: jamie
+Genomes in MABE2 are just repositories of mutable information.  The most basic kinds are vectors of some type, and (unlike in MABE), they aren't hidden inside Organisms which manage everything about them, but they are actively added to them and manage themselves.  This means that to mutate a Genome, any Genome, the Organism just tells the Genome to mutate itself and the Genome does.  No longer will the Organism run all the mutation functions itself, now it just delegates.  This, critically, means that Organisms now no longer need to know what kinds of Genomes they have at all.
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:28]: change unlike in MABE to unlike MABE to make this read better

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+          max_org = &org;
+        }
+      }
+
+      std::cout << "Max " << fitness_trait << " = " << max_fitness << std::endl;
+    }
+        
+```
+        
+The important thing here (and the thing which you will likely need to edit for any custom experiment is this bit of code:
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:193]: add ")"

Comment (AriaKillebrewBruehl) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:193]: explain why you would need to edit/how you would edit this bit  of code

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+## Tutorial - Creating and running a custom experiment using DynamicOrg
+
+Aagos (Auto-Adaptive Genetic Organization System), is a digital evolution experiment designed to show how high and low environmental change promote the evolution of segregated or non-segregated genes. It consists of N K long genes in a circular genome that can undergo insertion/deletion/in-place mutations, as well as gene move mutations. When a gene move mutation occurs, the gene randomly picks a new starting position in the genome. There is a maximum and minimum length the genome can grow/shrink to, and fitness is calculated by comparing the N genes to a set of gene targets that act as the environment. As a demonstration of DynamicOrg and its usefulness, we will show the steps involved in implementing Aagos using DynamicOrg.
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:216]: change "It consists of N K long" to "N K-long" to make this clearer (if its the case that there are N things that have length K)

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+  
+Below is a list of possible steps a user could take when implementing a custom experiment. For Aagos, only steps 1, 2 and 3 are required.
+  1. Write the Evaluator. (almost always necessary)
+      - Set up the config.
+      - Write the fitness calculator in OnUpdate from the emp::DataMap the Organism provides after org.GenerateOutput().
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:221]: add code formatting for org.GenerateOutput() etc. if possible

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+## Acknowledgements
+
+Mentors: Dr. Charles Ofria & Cliff Bohm
+Participants: Mitchell Johnson & Jamie Schmidt
+
+Charles Ofria and Cliff Bohm conceived and planned the idea of DynamicOrg. They both then contributed to the design and implementation of DynamicOrg. The majority of implementation of DynamicOrg was created by Jamie Schmidt with help from Mitchell Johnson.
+
+Jamie Schmidt wrote the sections on Brains, Genomes, Evaluators, and Introduction. Mitchell Johnson wrote the sections on Tutorial using DynamicOrg, and Conclusion. Jamie Schmidt and Mitchell Johnson worked together on sections Future plans, and DynamicOrg.
+
+Thank you to our amazing mentors!
```
Comment (AriaKillebrewBruehl) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:296]: I think there's a boilerplate disclaimer that needs to go at the end of this post 
"This work is supported through Active LENS: Learning Evolution and the Nature of Science using Evolution in Action (NSF IUSE #1432563). Any opinions, findings, and conclusions or recommendations expressed in this material are those of the author(s) and do not necessarily reflect the views of the National Science Foundation."

Comment (AriaKillebrewBruehl) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:296]: Good post overall! Very thorough explanation of what you did this summer! The code snippets were very helpful for understanding how things were implemented. It would be nice if you could all emojis, bolding, or italisizing to emphasize different parts of this post. Right now it is hard to tell what some of the main points are and that adding that formatting in would help!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+3. An Organism needs to return some kind of object which consists of all the data the Evaluator needs. 
+
+Of course, different Organisms can handle things differently, so for example an Organism in MABE (call it a BitOrg) could consist entirely of one TypedGenome<bool>, which is a Genome that is composed entirely of bits. It can copy that whenever it reproduces, it can flip some random number of bits every time it mutates, and it can simply return the bits of that Genome in a BitVector (a specialized vector of bits which Empirical uses in place of std::BitMaps and std::vector<bool>s).
+  
+But what's a DynamicOrg?  Well, a DynamicOrg is an Organism that is nearly totally built by the config file. Unlike our previous example, where the BitOrg knew it had a Genome of bits and how to mutate and return it, a DynamicOrg can't.  Ideally, you'd be able to specify what Genomes and Brains were in this DynamicOrg, and that would be it. Specifying Genomes is relatively easy, since they're all completely self-contained, but the Brains pose a more significant challenge, which is why we'll be discussing them in the section after this.
```
Comment (lemniscate8) [inline blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:81]: Why can't the ideal be achieved? What restrictions? I would move this to the other section if that's where this is actually covered, otherwise the reader is left hanging.

Also next paragraph starts with opposite tone making the case for DynamicOrg being less restrictive so it's odd to first mention there were significant challenges that will be addressed later.

**Review by lemniscate8 — CHANGES_REQUESTED** [review]
Really cool to learn a bit about how MABE2 works. It sounds really great how much you've streamlined the process for building experiments!
Some global things:
Check word choice in some places since some of your clarifications mutated into reiterations and some sentences may be on the wordy side.
I'd recommend using psuedo code for large sections or describe what the code is doing since some of it was pretty hard to parse for someone unfamiliar with the codebase.
Also Matthew recommended putting each sentence on a new line since and that's helped me a lot with version control/feedback since you can target individual sentences.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 11 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:11]: done, thank you!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 12 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:17]: good suggestions, done!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 14 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:193]: done

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 15 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:216]: That did look and sound a bit confusing, thank you

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 16 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:221]: done

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 17 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:397]: Thank you for the helpful suggestions!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+
+![MABE structure](https://github.com/Hintzelab/MABE/wiki/images/MABE_Overview.png)
+
+and, though powerful, could of course be improved.  The creation of the 2.0 version of MABE, called MABE2, was the focus of our project.  Our main focus was on refining the underlying architecture of MABE to be more end user-friendly, particularly in designing custom experiments.  The culmination of our efforts was in the creation of DynamicOrg, an Organism which was near-completely abstracted to allow for any combination of Genomes and Brains, as assigned through the config file.
+
+Along with the abstraction of organisms, we were also able to abstract away the inner workings of genomes and brains into a TypedGenome<type> and a BasicBrain. All of these abstractions remove many technical pre-requisites that users were required to have before using MABE2. Because of this, not only are far fewer steps involved when setting up an experiment, but it also takes much less time, since the majority of designing a simple experiment should be simple plug-and-play.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:19]: good suggestion, i added some inline code whenever code was used

[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:19]: I also added descriptions under especially confusing bits of code, thank you!

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+  
+For comparison, this is the structure of MABE2 when using a DynamicOrg:
+  
+![MABE2 structure](https://i.imgur.com/TCklKle.png)
+  
+To explain exactly what's going on here, we're going to go through Genomes, DynamicOrg, Brains, and then Evaluators.
+
+## Genomes / author: jamie
+Genomes in MABE2 are just repositories of mutable information.  The most basic kinds are vectors of some type, and (unlike in MABE), they aren't hidden inside Organisms which manage everything about them, but they are actively added to them and manage themselves.  This means that to mutate a Genome, any Genome, the Organism just tells the Genome to mutate itself and the Genome does.  No longer will the Organism run all the mutation functions itself, now it just delegates.  This, critically, means that Organisms now no longer need to know what kinds of Genomes they have at all.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:29]: good catch, its been reworded to be less confusing and more specific.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 21]
```diff
@@ -0,0 +1,397 @@
...
+
+Genomes in MABE2 are also accessed differently -- regardless of what type the Genome is, it can be read from and written to as if it were made of doubles, ints, bytes, or bits.  They can also be read from and written to in bulk, and they can be "scaled", where what you see is automatically scaled to whatever the Genome's actual range of values is.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:31]: Changed the wording to make it more clear. read/written to "in bulk" means that multiple values or positions in a Genome can be read from/written to in one function call.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 22]
```diff
@@ -0,0 +1,397 @@
...
+        mut_sites.Set(pos); // sets each site of mutation to 1
+      }
+      data ^= mut_sites; // computes an exclusive or with the vector of bits and the mutation position vector
+
+      return num_muts;
+    }
+  
+```  
+
+Naturally, there are many more ways to mutate a Genome, particuarly if the Genome is a vector of integers or doubles.  If you want to make a Genome that mutates in a particular fashion, this is the way you'll have to go about it.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:72]: Jamie can answer this better than me, but I reworded it to make it less confusing.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 23]
```diff
@@ -0,0 +1,397 @@
...
+
+Genomes in MABE2 are also accessed differently -- regardless of what type the Genome is, it can be read from and written to as if it were made of doubles, ints, bytes, or bits.  They can also be read from and written to in bulk, and they can be "scaled", where what you see is automatically scaled to whatever the Genome's actual range of values is.  
+
+Moreover, Genomes are often handled through the Genome::Head class, which allows for easy and abstract use of Genomes.  Heads move along a Genome, and through a Head, any Genome operation can be accomplished, including searching (for values, ranges, multiple values, and ranges of multiple values) and all aforementioned reads and writes.  
+
+This means that it is often beneficial to pass Genomes as Heads if you want to be as generic as possible while still allowing full functionality of the Genome.
+  
+The most important part of a Genome is the mutate function.  Here's an example of a very simple mutate function for a vector of bits:
+  
+```cpp
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:39]: I added a description to help the reader along with the following code. This kind of genome of 1s and 0s is standard in many of the experiments run in MABE. But theres no motivation from nature here, just random mutations that occur on rebirths between organisms.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 24]
```diff
@@ -0,0 +1,397 @@
...
+In the config file, this means all a DynamicOrg needs is a list of names of all the Genomes and Brains it will contain. So long as all these Genomes and Brains are properly named, this will work seamlessly. Each specific TypedGenome<> will also come with specific parameters that appear in the .mabe file. Parameters like mutation probabilities, max/min values, and initially starting randomized or not, allow very specific experiments to be created.
+        
+The most important part of DynamicOrg is GenerateOutput().  
+        
+ ```cpp
+        
+    void GenerateOutput() override {
+      emp::DataMap temp;
+  
+      map.Set<emp::vector<emp::Ptr<Genome>>>("genomes", genome_vec);
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:96]: typo

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 25]
```diff
@@ -0,0 +1,397 @@
...
+        brain_vec[i]->process(map);
+      }
+
+      SetVar<emp::DataMap>(SharedData().output_name, temp);
+    }
+        
+        
+```
+        
+Though it may look very simple, in fact it tells you exactly what goes on in Brains and Evaluators.  Essentially, the Brains get sent a DataMap of the Genomes, the DynamicOrg's own DataMap, an output DataMap, and (not seen here) the output name.  Then, each Brain builds itself (which most Brains don't do) and then processes the data.  Each Brain writes its outputs to the output DataMap, which each Brain can also read from as well, allowing for communication of a kind.  Critically, this output DataMap is the only thing the Evaluator gets, and so is the only part of the DynamicOrg which it can see.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:110]: done, good idea

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 26]
```diff
@@ -0,0 +1,397 @@
...
+  
+## Brains / author: jamie
+
+Brains are the processors of all the Genomes in an Organism.  They are only ever accessed through their process function, which takes an emp::DataMap and then processes all the relevant information.  In DynamicOrg in particular, they are all given a DataMap containing the Genomes in the DynamicOrg, and as each Brain will know which Genomes matter for its calculations, it will then access those Genomes, process them, and write its output to the DataMap.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:114]: Added a line that clarifies

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md [hunk 27]
```diff
@@ -0,0 +1,397 @@
...
+  
+The Brain will be configured with the names of the Genomes it works with, along with the name of its output.  This means that you can make identical Brains which still deal with different Genomes and write their outputs to different addresses in the DataMap.  
+  
+Many Brains, such as Markov Brains, are built from Genomes themselves, and this also occurs in the process function.  This allows for Brains that evolve over time, and the Evaluator can either receive the Brain itself or whatever decision-making process it has built from the Genomes the Brain was given.
+        
+The most important part of a Brain is the process() function.
```
[the remark this answers is not in the document]
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:120]: reworded

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 18 above)
Comment (mitchelljohnson3) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md:83]: The ideal was achieved, it was just a weirdly worded line.

File: blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (schmi710) [reply blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md]: I elaborated on that further in my earliest draft, but I ended up scrapping that section since it got too long and too off-topic, not to mention technical beyond the scope of the blog post.  Mitchel did a great job in clarifying it.
